modules/http/http-loadbalancer.c

```diff
@@ -0,0 +1,285 @@
+          remainder--;
+        }
+      msg_trace("Setting maximum number of workers for HTTP destination",
+                evt_tag_str("url", target->url),
+                evt_tag_int("max_clients", target->max_clients));
+    }
+}
+
+static void
+_switch_target(HTTPLoadBalancer *self, HTTPLoadBalancerClient *lbc, HTTPLoadBalancerTarget *new_target)
+{
+  /* self->lock must be held */
+  if (lbc->target != new_target)
+    {
+      if (lbc->target)
+        lbc->target->number_of_clients--;
+
+      new_target->number_of_clients++;
+      lbc->target = new_target;
+    }
+}
+
+static HTTPLoadBalancerTarget *
+_get_least_recently_tried_failing_target(HTTPLoadBalancer *self)
+{
+  time_t lru = 0;
+  gint lru_index = -1;
+
+  for (gint i = 0; i < self->num_targets; i++)
+    {
+      HTTPLoadBalancerTarget *target = &self->targets[i];
+
+      if (target->state != HTTP_TARGET_FAILED)
+        continue;
+
+      if (lru_index < 0 || lru > target->last_failure_time)
+        {
+          lru = target->last_failure_time;
+          lru_index = i;
+        }
+    }
+  return &self->targets[lru_index >= 0 ? lru_index : 0];
+}
+
+static HTTPLoadBalancerTarget *
+_recover_a_failed_target(HTTPLoadBalancer *self)
+{
+  self->last_recovery_attempt = time(NULL);
+  /* allocation failed, try the target that failed the least recently. */
+  return _get_least_recently_tried_failing_target(self);
+}
+
+static gboolean
+_check_recovery(HTTPLoadBalancer *self, HTTPLoadBalancerClient *lbc, HTTPLoadBalancerTarget **new_target)
+{
+  if (self->num_failed_targets > 0)
+    {
+      time_t now = time(NULL);
+
+      if (self->last_recovery_attempt == 0)
+        self->last_recovery_attempt = now;
+
+      if (now - self->last_recovery_attempt >= self->recovery_timeout)
+        {
+          *new_target = _recover_a_failed_target(self);
+          return TRUE;
+        }
+    }
+  return FALSE;
+}
+
+static HTTPLoadBalancerTarget *
+_locate_target(HTTPLoadBalancer *self, HTTPLoadBalancerClient *lbc)
+{
+  gint start_index = lbc->target
+                     ? (lbc->target->index + 1) % self->num_targets
+                     : 0;
+  for (gint i = 0; i < self->num_targets; i++)
+    {
+      HTTPLoadBalancerTarget *target = &self->targets[i + start_index];
+
+      if (target->state == HTTP_TARGET_OPERATIONAL &&
+          target->number_of_clients < target->max_clients)
+        return target;
+    }
+
+  return _recover_a_failed_target(self);
+}
+
+static gboolean
+_check_rebalance(HTTPLoadBalancer *self, HTTPLoadBalancerClient *lbc, HTTPLoadBalancerTarget **new_target)
+{
+  /* Are we misbalanced? */
+  if (lbc->target == NULL ||
+      lbc->target->state != HTTP_TARGET_OPERATIONAL ||
+      lbc->target->number_of_clients > lbc->target->max_clients)
+    {
+      *new_target = _locate_target(self, lbc);
+      return TRUE;
+    }
+  return FALSE;
+}
+
+HTTPLoadBalancerTarget *
+http_load_balancer_choose_target(HTTPLoadBalancer *self, HTTPLoadBalancerClient *lbc)
+{
```

Comment:
1.What if we wouldn't choose a new target when the number of workers less then the number of targets(`num_clients` < num_targets)?